docs: gRPC advanced guides (#3034)
bentoml/BentoML: docs/source/guides/grpc.rst

- Streaming support (currently on our roadmap).

This guide will also walk your through some of the strengths and weaknesses of serving with gRPC, as well as
recommendation on scenarios where gRPC might be a good fit.

:bdg-info:`Requirements:` This guide assumes that you have basic knowledge of gRPC and protobuf. If you aren't
familar with gRPC, you can start with gRPC `quick start guide <https://grpc.io/docs/languages/python/quickstart/>`_.

.. seealso::

   For quick introduction to serving with gRPC, see :ref:`Intro to BentoML <tutorial:Tutorial: Intro to BentoML>`

Get started with gRPC in BentoML
--------------------------------

We will be using the example from :ref:`the quickstart<tutorial:Tutorial: Intro to BentoML>` to
demonstrate BentoML capabilities with gRPC.

Requirements
~~~~~~~~~~~~

Install BentoML with gRPC support with :pypi:`pip`:

.. code-block:: bash

   » pip install "bentoml[grpc]"

Thats it! You can now serving your Bento with gRPC via :ref:`bentoml serve-grpc <reference/cli:serve-grpc>` without having to modify your current service definition 😃.

.. code-block:: bash

   » bentoml serve-grpc iris_classifier:latest --production

> Interact with your gRPC BentoService

Author: I changed it to `Using your gRPC BentoService`?

      » bentoml serve-grpc iris_classifier:latest --production --enable-reflection

   .. include:: ./snippets/grpc/docs/grpc_tools.rst

   Open a different terminal and use one of the following:

2. Use one of the below :ref:`client implementations <guides/grpc:Client Implementation>` to send test requests to your BentoService.

.. _workspace: https://bazel.build/concepts/build-ref

.. |workspace| replace:: ``WORKSPACE``

.. _build: https://bazel.build/concepts/build-files

.. |build| replace:: ``BUILD``

Client Implementation
~~~~~~~~~~~~~~~~~~~~~

.. note::

   All of the following client implementations are :github:`available on GitHub <bentoml/BentoML/tree/main/docs/source/guides/snippets/grpc/>`.

:raw-html:`<br />`

From another terminal, use one of the following client implementation to send request to the
gRPC server:

.. note::

   There are many different ways to generate stubs and implement your own gRPC clients.
   Please feel free to use any workflow you like to generate stubs. 

   With that being said, we recommend users to use either of the following two approaches:

   - Using `bazel <bazel.build>`_ to manage and isolate dependencies (recommended)

>    - A manual approach using ``protoc`` its language-specific plugins

Reviewer: Not sure what this is recommending. Is it about how to generate the stubs from user's code base?

Author: I should just remove this section.

  then reshape the given payload into expected shape.

  Note that this value will always takes precendence over the ``shape`` field in the :ref:`bentoml.io.NumpyNdarray <reference/api_io_descriptors:NumPy \`\`ndarray\`\`>` descriptor,
  meaning the array will be reshaped to this value first if given. Refer to :meth:`bentoml.io.NumpyNdarray.from_proto` for implementation details.

* `string_values`, `float_values`, `double_values`, `bool_values`, `int32_values`, `int64_values`, `uint32_values`, `unit64_values`

  Each of the fields is a `list` of the corresponding data type. The list is a flattened array, and will be reconstructed
  alongside with ``shape`` field to the original payload.

  Per request sent, one message should only contain **ONE** of the aforementioned fields.

  The interaction among the above fields and ``dtype`` are as follows:

  - if ``dtype`` is not present in the message:
      * All of the fields are empty, then we return a ``np.empty``.
      * We will loop through all of the provided fields, and only allows one field per message.

>         If here are more than two fields (i.e. ``string_values`` and ``float_values``), then we will raise an error, as we don't know how to deserialize the data.

Reviewer: more than one field?

  - otherwise:

>       * We will use the provided dtype-to-field maps to get the data from the given message.

Reviewer: Link from dtype-to-field. Also, s/maps/map

  For example, if ``dtype`` is ``DTYPE_FLOAT``, then the payload expects to have ``float_values`` field.

.. grid:: 2

    .. grid-item-card::  ``Python API``

      .. code-block:: python

         NumpyNdarray.from_sample(
            np.array([[5.4, 3.4, 1.5, 0.4]])
         )

    .. grid-item-card::  ``pb.NDArray``

      .. code-block:: none

         ndarray {
           dtype: DTYPE_FLOAT
           shape: 1
           shape: 4
           float_values: 5.4
           float_values: 3.4
           float_values: 1.5
           float_values: 0.4
         }

:bdg-primary:`API reference:` :meth:`bentoml.io.NumpyNdarray.from_proto`

:raw-html:`<br />`

Tabular data representation via ``DataFrame``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

:bdg-info:`Description:` ``DataFrame`` represents any tabular data type. Currently we only support the columns orientation

> since it is turned out to be the only orientation that we can preserve ordering.

Reviewer: ```suggestion
since it is it is best for preserving the input order.
```

> This means that every other orientation is currently not yet supported.

Reviewer: Kind of redundant.

its coresponding :class:`bentoml.io.IODescriptor`

.. grid:: 2

    .. grid-item-card::  ``Python API``

      .. code-block:: python

         Multipart(
            meta=Text(),
            arr=NumpyNdarray(
               dtype=np.float16,
               shape=[2,2]
            )
         )

    .. grid-item-card::  ``pb.Multipart``

      .. code-block:: none

         multipart {
            fields {
               key: "arr"
               value {
                  ndarray {
                  dtype: DTYPE_FLOAT
                  shape: 2
                  shape: 2
                  float_values: 1.0
                  float_values: 2.0
                  float_values: 3.0
                  float_values: 4.0
                  }
               }
            }
            fields {
               key: "meta"
               value {
                  text {
                  value: "nlp"
                  }
               }
            }
         }

:bdg-primary:`API reference:` :meth:`bentoml.io.Multipart.from_proto`

Compact data format via ``serialized_bytes``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> We introduce the field ``serialized_bytes`` to both ``Request`` and ``Response`` such
> that the payload is serialized with BentoML's internal serialization format.
> 
> This is useful to when we want to send a large amount of data on the wire.
> However, as mentioned above, this is an internal serialization format and thus not
> **recommended** for use by users.

Reviewer: ```suggestion
The ``serialized_bytes`` field in both ``Request`` and ``Response``  is reserved for pre-established protocol encoding between client and server. BentoML leverages the field to improve serialization performance between BentoML client and server. Thus the field is not **recommended** for use directly.
```

Mounting Servicer
~~~~~~~~~~~~~~~~~

With support for :ref:`multiplexing <guides/grpc:Demystifying the misconception of gRPC vs. REST>`
to eliminate :wiki:`head-of-line blocking <Head-of-line_blocking>`,
gPRC enables us to mount additional custom servicess alongside with BentoService,

> and serve them under the same port.

Reviewer: Mounting services is independent of multiplexing and head of line blocking. Only that gRPC multiplexing makes it nicer so that one service does not block another service.

Also, there is a typo in "servicess".

Author: Isn't that the purpose of service multiplexing? is that multiple services can be served under the same port?

Reviewer: You can technically do that in http/1 without multiplexing, like uvicorn.

Author: but i guess it is not support ootb

.. code-block:: bash

   » bentoml serve-grpc service.py:svc --reload --enable-reflection

Now your ``RouteGuide`` service can also be accessed through ``localhost:3000``.

.. note::

   ``service_names`` is **REQUIRED** here, as this will be used for :github:`server reflection <grpc/grpc/blob/master/doc/server-reflection.md>`
   when ``--enable-reflection`` is passed to ``bentoml serve-grpc``.

Mounting gRPC Interceptors
~~~~~~~~~~~~~~~~~~~~~~~~~~

Inteceptors are a component of gRPC that allows us to intercept and interact with the
proto message and service context either before - or after - the actual RPC call was
sent/received by client/server.

Interceptors to gRPC is what middleware is to HTTP. The most common use-case for interceptors
are authentication, :ref:`tracing <guides/tracing:Tracing>`, access logs, and more.

BentoML comes with a sets of built-in *async interceptors* to provide support for access logs,
`OpenTelemetry <https://opentelemetry.io/>`_, and `Prometheus <https://prometheus.io/>`_.

The following diagrams demonstrates the flow of a gRPC request from client to server:

.. image:: /_static/img/interceptor-flow.svg

>    :alt: Interceptor Flow

Reviewer: The diagram looks great. Only the text looks a bit small on my screen. Could you please make the text larger?

Since interceptors are executed in the order they are added, users interceptors will be executed after the built-in interceptors.

> Users interceptors should be **READ-ONLY**, which means it shouldn't modify the state or content of the incoming ``Request``.

Reviewer: Read-only is a bit misleading (we gave an example below modifying the request with additional metadata. :)). Maybe we can say user interceptors should preserve the existing headers and data of the request.

Author: make sense.

true.

gRPC is built on top of HTTP/2, and it addresses some of the shortcomings of HTTP/1.1,
such as :wiki:`head-of-line blocking <Head-of-line_blocking>`, and :wiki:`HTTP pipelining <HTTP_pipelining>`.
However, gRPC is not a replacement for REST, and indeed it is not a replacement for
model serving. gRPC comes with its own set of trade-offs, such as:

* **Limited browser support**: It is impossible to call a gRPC service directly from any
  browser. You will end up using tools such as :github:`gRPCUI <fullstorydev/grpcui>` in order to interact
  with your service, or having to go through the hassle of implementing a gRPC client in
  your language of choice.

* **Binary protocol format**: While :github:`Protobuf <protocolbuffers/protobuf>` is
  efficient to send and receive over the wire, it is not human-readable. This means
  additional toolin for debugging and analyzing protobuf messages are required.

* **Knowledge gap**: gRPC comes with its own concepts and learning curve, which requires
  teams to invest time in filling those knowledge gap to be effectively use gRPC. This
  often leads to a lot of friction and sometimes increase friction to the development
  agility.

> * **Lack of suport for additional content types**: gRPC depends on protobuf, its content

Reviewer: ```suggestion
* **Lack of support for additional content types**: gRPC depends on protobuf, its content
```

=================
Serving with gRPC
=================

This guide will demonstrate advanced features that BentoML offers for you to get started
with `gRPC <https://grpc.io/>`_:

- First-class support for :ref:`custom gRPC Servicer <guides/grpc:Mounting Servicer>`, :ref:`custom interceptors <guides/grpc:Mounting gRPC Interceptors>`, handlers.
- Seemlessly adding gRPC support to existing Bento.

This guide will also walk you through tradeoffs of serving with gRPC, as well as
recommendation on scenarios where gRPC might be a good fit.

:bdg-info:`Requirements:` This guide assumes that you have basic knowledge of gRPC and protobuf. If you aren't
familar with gRPC, you can start with gRPC `quick start guide <https://grpc.io/docs/languages/python/quickstart/>`_.

.. seealso::

   For quick introduction to serving with gRPC, see :ref:`Intro to BentoML <tutorial:Tutorial: Intro to BentoML>`

Get started with gRPC in BentoML
--------------------------------

We will be using the example from :ref:`the quickstart<tutorial:Tutorial: Intro to BentoML>` to
demonstrate BentoML capabilities with gRPC.

Requirements
~~~~~~~~~~~~

> BentoML supports for gRPC are introduced in version 1.0.7 and above.

Reviewer: ```suggestion
BentoML supports for gRPC are introduced in version 1.0.6 and above.
```
